File: android/src/main/java/com/audiocontext/nodes/AudioNode.kt
package com.audiocontext.nodes

import com.audiocontext.context.BaseAudioContext
import com.audiocontext.parameters.PlaybackParameters
import com.facebook.jni.HybridData


abstract class AudioNode(val context: BaseAudioContext) {
  open val numberOfInputs: Int = 0
    get() = field
  open val numberOfOutputs: Int = 0
    get() = field
  private val inputNodes = mutableListOf<AudioNode>()
  private val outputNodes = mutableListOf<AudioNode>()

  private val mHybridData: HybridData?;

  companion object {
    init {
      System.loadLibrary("react-native-audio-context")
    }
  }

  init {
    mHybridData = initHybrid()
  }

  external fun initHybrid(): HybridData?

  fun connect(node: AudioNode) {
    if(numberOfOutputs > outputNodes.size && node.numberOfInputs > node.inputNodes.size) {
      outputNodes.add(node)
      node.inputNodes.add(this)
    }
  }

  fun disconnect(node: AudioNode) {
    outputNodes.remove(node)
    node.inputNodes.remove(this)
  }

  open fun process(playbackParameters: PlaybackParameters) {
    outputNodes.forEach { it.process(playbackParameters) }
  }

  open fun cleanup() {
    outputNodes.forEach { it.cleanup() }
    inputNodes.clear()
    outputNodes.clear()
  }
}

File: ios/nodes/AudioNode/AudioNode.m
#import "AudioNode.h"
#import "AudioContext.h"

@implementation AudioNode

- (instancetype)initWithContext:(AudioContext *)context {
    if (self = [super init]) {
        _connectedNodes = [NSMutableArray array];
        _context = context;
    }

    return self;
}

- (void)cleanup {
    _context = nil;
    [_connectedNodes removeAllObjects];
    _connectedNodes = nil;
    
}

- (void)processWithParameters:(PlaybackParameters *)parameters; {
    for (AudioNode *node in _connectedNodes) {
        [node processWithParameters:parameters];
    }
}

- (void)connect:(AudioNode *)node {
    if (_numberOfOutputs > 0) {
        [_connectedNodes addObject:node];
    }
}

- (void)disconnect:(AudioNode *)node {
    NSUInteger index = [_connectedNodes indexOfObject:node];
    if (index != NSNotFound) {
        [_connectedNodes removeObjectAtIndex:index];
    }
}

- (int)getNumberOfInputs {
    return _numberOfInputs;
}

- (int)getNumberOfOutputs {
    return _numberOfOutputs;
}

@end
